### PR TITLE
DROOLS-3957 : Make verifier panel use docks

### PR DESCRIPTION
**Pull request by Rikkola**
https://issues.jboss.org/browse/DROOLS-3957

@gitgabrio @jomarko @Bonuseto 
Here is a change I did ages ago as a subtask, but would like to get the changes in for two reasons:
* This unifies the look & feel with other editors. The old implementation had the same panel, but in a "floating" screen.
* The ticket for 3 stage verification might come up again later.

Asked @gitgabrio since he has been following the dock changes.

https://github.com/kiegroup/kie-wb-common/pull/2784
https://github.com/kiegroup/drools-wb/pull/1201

### CODE REVIEW

**Comment by Rikkola** [comment]
Fixing test failures.

**Comment by jomarko** [comment]
@Rikkola Hello, thank you for the PR. As you know I have locally troubles to verify any PR related to verifier. Can you please provide detailed steps, how to build the ensemble?

File: kie-wb-common-services/kie-wb-common-verifier/kie-wb-common-verifier-reporting-client/pom.xml [hunk 1]
```diff
@@ -79,6 +79,10 @@
       <groupId>org.kie.workbench.widgets</groupId>
       <artifactId>kie-wb-decorated-grid-widget</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.kie.workbench.widgets</groupId>
+      <artifactId>kie-wb-common-ui</artifactId>
```
Comment (jomarko) [inline kie-wb-common-services/kie-wb-common-verifier/kie-wb-common-verifier-reporting-client/pom.xml]: Is this dependency really needed?

File: kie-wb-common-widgets/kie-wb-metadata-widget/src/main/java/org/kie/workbench/common/widgets/metadata/client/KieMultipleDocumentEditor.java [hunk 2]
```diff
@@ -264,6 +277,31 @@ public void registerDocument(final D document) {
...
+                                final IsWidget widget) {
+        placeManager.registerOnOpenCallback(placeRequest, () -> {
+            if (getDockPresenter(placeRequest).isPresent()) {
+                getDockPresenter(placeRequest).get().setView(widget);
+            }
+        });
+    }
+
+    private Optional<PlaceHolderBase> getDockPresenter(final PlaceRequest placeRequest) {
+        if (PlaceStatus.OPEN.equals(placeManager.getStatus(placeRequest))) {
```
Comment (jomarko) [inline kie-wb-common-widgets/kie-wb-metadata-widget/src/main/java/org/kie/workbench/common/widgets/metadata/client/KieMultipleDocumentEditor.java]: Could we use `Objects.equals` please?

**Comment by jomarko** [comment]
@Rikkola I am sorry, manual acceptance test [failed](https://issues.jboss.org/browse/DROOLS-3957?focusedCommentId=13764819&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-13764819).

**Comment by Rikkola** [comment]
jenkins please retest this

**Comment by jomarko** [comment]
@Rikkola hello, please what is status of this PR, is it ready for a re-review?

**Comment by Rikkola** [comment]
@jomarko ok, PRs updated and ready for recheck.

**Comment by jomarko** [comment]
@Rikkola there is a failing test `org.kie.workbench.common.widgets.metadata.client.KieMultipleDocumentEditorTest.registerDockWhenItExists`

**Comment by jomarko** [comment]
@Rikkola may I ask you to double check your rebase? TO me the build is downloading 7.25.0-SNAPSHOT

**Comment by Rikkola** [comment]
@jomarko Rebased, but the test is not yet fixed. Taking a look now.

**Comment by Rikkola** [comment]
@jomarko and now even the test is fixed.

File: kie-wb-common-widgets/kie-wb-metadata-widget/src/test/java/org/kie/workbench/common/widgets/metadata/client/KieMultipleDocumentEditorTest.java [hunk 3]
```diff
@@ -830,4 +842,58 @@ public void testOpenDocumentInEditor_OneDocumentNotAlreadyRegistered() {
...
+        doReturn(workbenchActivity).when(placeManager).getActivity(any());
+        final PlaceHolderBaseView placeHolderView = mock(PlaceHolderBaseView.class);
+        doReturn(placeHolderView).when(workbenchActivity).getWidget();
+        verify(placeManager).registerOnOpenCallback(any(),
+                                                    commandArgumentCaptor.capture());
+
+        commandArgumentCaptor.getValue().execute();
+
+        verify(placeHolderView, never()).setWidget(widget);
+    }
```
Comment (jomarko) [inline kie-wb-common-widgets/kie-wb-metadata-widget/src/test/java/org/kie/workbench/common/widgets/metadata/client/KieMultipleDocumentEditorTest.java]: Names of these two is a little bit confusing from my point of view. the only difference is in `editor.showDocks = true;` while test name mention `exist` vs `doesn't exist`. Shouldn't test names reflect rather `docks expanded` vs `docks collapsed` or `shown` vs `hidden` . Not author of the code, sorry if my question is wrong and names are correct.

**Comment by jomarko** [comment]
### Manual check
Passed

File: kie-wb-common-widgets/kie-wb-common-ui/src/main/java/org/kie/workbench/common/widgets/client/docks/PlaceHolderBase.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    @Inject
+    public void DockPlaceHolder(final PlaceHolderBaseView view) {
+        this.view = view;
+        this.view.setPresenter(this);
+    }
+
+    @WorkbenchPartTitle
+    public String getTitle() {
+        return "DockPlaceHolder"; // Never used.
```
[the remark this answers is not in the document]
Comment (Rikkola) [reply kie-wb-common-widgets/kie-wb-common-ui/src/main/java/org/kie/workbench/common/widgets/client/docks/PlaceHolderBase.java]: Yeah even when it is an annotation it needs to be there.

[the remark this answers is not in the document]
Comment (jomarko) [reply kie-wb-common-widgets/kie-wb-common-ui/src/main/java/org/kie/workbench/common/widgets/client/docks/PlaceHolderBase.java]: @Rikkola  ok, please have a look on the rest of the comments so we can merge

**Comment by jomarko** [comment]
jenkins please retest this

File: kie-wb-common-widgets/kie-wb-metadata-widget/src/main/java/org/kie/workbench/common/widgets/metadata/client/KieMultipleDocumentEditor.java [hunk 5]
```diff
@@ -266,6 +282,38 @@ public void registerDocument(final D document) {
...
+                                final IsWidget widget) {
+        placeManager.registerOnOpenCallback(placeRequest, () -> {
+            if (showDocks) {
+                setDockWidget(placeRequest, widget);
+            }
+        });
+    }
+
+    protected void setDockWidget(final PlaceRequest placeRequest,
+                               final IsWidget widget) {
```
[the remark this answers is not in the document]
Comment (Rikkola) [reply kie-wb-common-widgets/kie-wb-metadata-widget/src/main/java/org/kie/workbench/common/widgets/metadata/client/KieMultipleDocumentEditor.java]: Fixed.

File: kie-wb-common-widgets/kie-wb-metadata-widget/src/test/java/org/kie/workbench/common/widgets/metadata/client/KieMultipleDocumentEditorTest.java [hunk 6]
```diff
@@ -830,4 +842,58 @@ public void testOpenDocumentInEditor_OneDocumentNotAlreadyRegistered() {
...
+    public void registerDock() {
+        editor.registerDock("test", mock(IsWidget.class));
+
+        verify(placeManager).registerOnOpenCallback(placeRequestArgumentCaptor.capture(),
+                                                    any(Command.class));
+        final PlaceRequest placeRequest = placeRequestArgumentCaptor.getValue();
+        assertEquals("org.docks.PlaceHolder", placeRequest.getIdentifier());
+        final Map<String, String> parameters = placeRequest.getParameters();
+        assertEquals(1, parameters.size());
+        assertEquals("test", parameters.get("name"));
```
[the remark this answers is not in the document]
Comment (Rikkola) [reply kie-wb-common-widgets/kie-wb-metadata-widget/src/test/java/org/kie/workbench/common/widgets/metadata/client/KieMultipleDocumentEditorTest.java]: Used the higher one. Really nice.

File: kie-wb-common-widgets/kie-wb-metadata-widget/src/test/java/org/kie/workbench/common/widgets/metadata/client/KieMultipleDocumentEditorTest.java (hunk 3 above)
Comment (Rikkola) [reply kie-wb-common-widgets/kie-wb-metadata-widget/src/test/java/org/kie/workbench/common/widgets/metadata/client/KieMultipleDocumentEditorTest.java]: Renamed the methods.

File: kie-wb-common-widgets/kie-wb-metadata-widget/src/main/java/org/kie/workbench/common/widgets/metadata/client/KieMultipleDocumentEditor.java [hunk 7]
```diff
@@ -266,6 +282,39 @@ public void registerDocument(final D document) {
...
+
+    protected void setDockWidget(final PlaceRequest placeRequest,
+                                 final IsWidget widget) {
+        final Optional<PlaceHolderBase> dockPresenter = getDockPresenter(placeRequest);
+        if (dockPresenter.isPresent()) {
+            dockPresenter.get().setView(widget);
+        }
+    }
+
+    private Optional<PlaceHolderBase> getDockPresenter(final PlaceRequest placeRequest) {
```
Comment (gitgabrio) [inline kie-wb-common-widgets/kie-wb-metadata-widget/src/main/java/org/kie/workbench/common/widgets/metadata/client/KieMultipleDocumentEditor.java]: @Rikkola 
Could you pls rename the method to reflect what is actually returned?

File: kie-wb-common-services/kie-wb-common-verifier/kie-wb-common-verifier-reporting-client/src/main/java/org/kie/workbench/common/services/verifier/reporting/client/panel/AnalysisDockPlaceHolder.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+package org.kie.workbench.common.services.verifier.reporting.client.panel;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.kie.workbench.common.widgets.client.docks.PlaceHolderBase;
+import org.uberfire.client.annotations.WorkbenchScreen;
+
+@ApplicationScoped
+@WorkbenchScreen(identifier = AnalysisDockPlaceHolder.IDENTIFIER)
+public class AnalysisDockPlaceHolder
```
Comment (gitgabrio) [inline kie-wb-common-services/kie-wb-common-verifier/kie-wb-common-verifier-reporting-client/src/main/java/org/kie/workbench/common/services/verifier/reporting/client/panel/AnalysisDockPlaceHolder.java]: @Rikkola 
Hi. In the modified files I read `PlaceHolderBase`, `DockPlaceHolder` and this `AnalysisDockPlaceHolder`. Following the names, I would think that `AnalysisDockPlaceHolder` is a specialization of `DockPlaceHolder`. If it is not so, maybe something could be renamed.

**Comment by Rikkola** [comment]
@gitgabrio Some enhancements. I hope you like them.

**Comment by Rikkola** [comment]
Hopefully my latest changes increase the test coverage enough. Not sure if there is anything left to test that makes sense to test.

**Comment by Rikkola** [comment]
66.7% Condition Coverage on New Code (is less than 80%)

**Comment by jomarko** [comment]
@Rikkola hello, I think the only method you need to increase coverage is `KieMultipleDocumentEditor#getDockPlaceHolder`, WDYT?

**Comment by Rikkola** [comment]
@jomarko Trying that, now all the paths for `getDockPlaceHoder `should be tested.

**Comment by jomarko** [comment]
@Rikkola as you may see, it is approved by me already.

**Comment by jomarko** [comment]
@gitgabrio just friendly reminder, could you please have a look on this ensemble? Seems @Rikkola has resolved your comments.

**Comment by Rikkola** [comment]
@gitgabrio The service used by this should be runnable in Kogito in one way or another. It is all client side code.

@kiegroup/gatekeepers Can I get a merge please.

**Pull request merged**